File: app/src/main/java/life/plank/juna/zone/view/adapter/FixtureDateAdapter.java
package life.plank.juna.zone.view.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.lang.ref.WeakReference;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import life.plank.juna.zone.R;
import life.plank.juna.zone.data.model.FixtureByDate;
import life.plank.juna.zone.util.BaseRecyclerView;
import life.plank.juna.zone.view.activity.base.BaseLeagueActivity;

import static life.plank.juna.zone.util.DateUtil.getDateHeader;

public class FixtureDateAdapter extends BaseRecyclerView.Adapter<FixtureDateAdapter.FixtureDateViewHolder> {

    private List<FixtureByDate> fixtureByDateList;
    private BaseLeagueActivity activity;

    FixtureDateAdapter(List<FixtureByDate> fixtureByDateList, BaseLeagueActivity activity) {
        this.fixtureByDateList = fixtureByDateList;
        this.activity = activity;
    }

    @Override
    public FixtureDateViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new FixtureDateViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_fixture_date, parent, false), this);
    }

    @Override
    public int getItemCount() {
        return fixtureByDateList.size();
    }

    static class FixtureDateViewHolder extends BaseRecyclerView.ViewHolder {

        @BindView(R.id.date_time)
        TextView dateTime;
        @BindView(R.id.fixtures_date_list)
        RecyclerView recyclerView;

        private final WeakReference<FixtureDateAdapter> ref;

        FixtureDateViewHolder(View itemView, FixtureDateAdapter fixtureDateAdapter) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            ref = new WeakReference<>(fixtureDateAdapter);
        }

        @Override
        public void bind() {
            FixtureByDate fixtureByDate = ref.get().fixtureByDateList.get(getAdapterPosition());
            dateTime.setText(getDateHeader(fixtureByDate.getDate()));
            recyclerView.setAdapter(new FixtureAdapter(fixtureByDate.getFixtures(), ref.get().activity));
        }
    }
}

File: app/src/main/java/life/plank/juna/zone/ui/football/adapter/fixture/FixtureDateAdapter.kt
package life.plank.juna.zone.ui.football.adapter.fixture

import android.view.*
import androidx.recyclerview.widget.RecyclerView
import com.prembros.facilis.util.isNullOrEmpty
import kotlinx.android.synthetic.main.item_fixture_date.view.*
import life.plank.juna.zone.R
import life.plank.juna.zone.data.model.football.FixtureByDate
import life.plank.juna.zone.util.time.DateUtil.getDateHeader
import life.plank.juna.zone.ui.football.LeagueContainer

class FixtureDateAdapter(
        private val fixtureByDateList: List<FixtureByDate>,
        private val leagueContainer: LeagueContainer
) : RecyclerView.Adapter<FixtureDateAdapter.FixtureDateViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FixtureDateViewHolder =
            FixtureDateViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_fixture_date, parent, false))

    override fun onBindViewHolder(holder: FixtureDateViewHolder, position: Int) {
        holder.itemView.run {
            val (date, fixtures) = fixtureByDateList[position]
            date_time.text = if (!isNullOrEmpty(fixtures))
                getDateHeader(fixtures[0].matchStartTime)
            else
                getDateHeader(date)
            fixtures_date_list.adapter = FixtureAdapter(fixtures, leagueContainer)
        }
    }

    override fun getItemCount(): Int = fixtureByDateList.size

    class FixtureDateViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
}
